cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/service/DraftHandler.java

```diff
@@ -0,0 +1,215 @@
+    });
+  }
+
+  /**
+   * Creates or updates a draft
+   */
+  @PUT
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void putDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                       @PathParam("context") String namespaceName,
+                       @PathParam("draft") String draftId) {
+
+    respond(namespaceName, responder, (namespace) -> {
+
+      String requestStr = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+      DraftStoreRequest<ETLConfig> draftStoreRequest = deserializeDraftStoreRequest(requestStr);
+
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+      draftService.writeDraft(id, draftStoreRequest);
+
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  /**
+   * Deletes a draft
+   */
+  @DELETE
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void deleteDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                          @PathParam("context") String namespaceName,
+                          @PathParam("draft") String draftId) {
+    respond(namespaceName, responder, (namespace) -> {
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+
+      draftService.deleteDraft(id);
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  /**
+   * Utility method to correct deserialize the config field in the {@link DraftStoreRequest} object
+   * @param jsonStr the json string representing the DraftStoreRequest
+   * @return {@link DraftStoreRequest} object
+   */
+  private DraftStoreRequest<ETLConfig> deserializeDraftStoreRequest(String jsonStr) {
+    try {
+      DraftStoreRequest<ETLConfig> request = GSON
+        .fromJson(jsonStr, new TypeToken<DraftStoreRequest<ETLConfig>>() { }.getType());
+
+      if (request.getArtifact() == null) {
+        throw new IllegalArgumentException("artifact is null");
+      }
+
+      if (StudioUtil.isBatchPipeline(request.getArtifact())) {
+        return GSON.fromJson(jsonStr, new TypeToken<DraftStoreRequest<ETLBatchConfig>>() { }.getType());
+      } else if (StudioUtil.isStreamingPipeline(request.getArtifact())) {
+        return GSON.fromJson(jsonStr, new TypeToken<DraftStoreRequest<DataStreamsConfig>>() { }.getType());
+      } else {
+        throw new IllegalArgumentException(String
+                                             .format("artifact '%s' is not supported, valid options are: '%s' or '%s'",
+                                                     request.getArtifact().getName(), StudioUtil.ARTIFACT_BATCH_NAME,
+                                                     StudioUtil.ARTIFACT_STREAMING_NAME));
+      }
+    } catch (JsonSyntaxException e) {
+      throw new IllegalArgumentException("Unable to decode request body: " + e.getMessage());
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Invalid config: " + e.getMessage());
```

Comment:
So this is catching the one thrown in line 164? Seems confusing. Also, generally, when wrapping an exception, always carry the origin exception as the cause, otherwise it will make the stacktrace not debuggable.

Reply:
Removed the catch for IllegalArguementException and passed on the cause for JsonSyntaxException